numpy/core/src/multiarray/item_selection.c

```diff
@@ -2096,10 +2103,13 @@ PyArray_CountNonzero(PyArrayObject *self)
             if (nonzero(data, self)) {
                 ++nonzero_count;
             }
+            if (needs_api && PyErr_Occurred()) {
+                return -1;
+            }
             data += stride;
         }
 
-        return nonzero_count;
+        return PyErr_Occurred() ? -1 : nonzero_count;
```

Comment:
In `lowlevel_strided_loops.c.src` I see a couple places which break inner loops on `(needs_api && PyErr_Occurred())`. So I do think we need to catch errors mid loop. Furthermore, as noted in another comment above, if we don't break the loop we are likely to go back into python code on the next looparound with an error set.

Also, to restate my last comment, currently we do not actually need to catch anything if `!needs_api`, but perhaps in the future we might.

And @seberg, nonzero can't return special values because it has to return `npy_bool`, which should only really be 0 or 1, even if technically that is a char.